File: walletsdk/src/main/java/leaf/prod/walletsdk/model/sign/Bitstream.java
package leaf.prod.walletsdk.model.sign;

import java.math.BigInteger;

import org.apache.commons.lang3.StringUtils;
import org.web3j.utils.Numeric;

import lombok.Data;
import lombok.val;

/**
 * Created with IntelliJ IDEA.
 * User: kenshin wangchen@loopring.org
 * Time: 2019-04-01 3:32 PM
 * Cooperation: loopring.org 路印协议基金会
 */

@Data
public class Bitstream {

    private String data;

    private static final Integer ADDRESS_LENGTH = 20;

    private static final BigInteger UINT256MAX = new BigInteger(StringUtils.repeat("f", 64), 16);

    public String getData() {
        return StringUtils.isEmpty(data) ? "0x0" : "0x" + data;
    }

    public byte[] getBytes() {
        return Numeric.hexStringToByteArray(data);
    }

    public Integer getLength() {
        return data.length() / 2;
    }

    public Integer addAddress(String address, Boolean forceAppend) {
        return addAddress(address, ADDRESS_LENGTH, forceAppend);
    }

    private Integer addAddress(String address, Integer numBytes, Boolean forceAppend) {
        String value = StringUtils.isEmpty(address) ? "0" : address;
        return insert(Numeric.toHexStringNoPrefixZeroPadded(Numeric.toBigInt(value), numBytes * 2), forceAppend);
    }

    public Integer addUint16(BigInteger num, Boolean forceAppend) {
        return addBigInt(num, 2, forceAppend);
    }

    public Integer addInt16(BigInteger num, Boolean forceAppend) {
        if (num.signum() != -1) {
            return addBigInt(num, 2, forceAppend);
        } else {
            BigInteger negative = num.add(UINT256MAX).add(BigInteger.ONE);
            String int16Str = negative.toString(16).substring(60, 64);
            return addHex(int16Str, forceAppend);
        }
    }

    public Integer addUint32(BigInteger num, Boolean forceAppend) {
        return addBigInt(num, 4, forceAppend);
    }

    public Integer addUint(BigInteger num, Boolean forceAppend) {
        return addBigInt(num, 32, forceAppend);
    }

    public Integer addNumber(BigInteger num, Integer numBytes, Boolean forceAppend) {
        return addBigInt(num, numBytes, forceAppend);
    }

    public Integer addBoolean(Boolean b, Boolean forceAppend) {
        BigInteger i = b ? BigInteger.ONE : BigInteger.ZERO;
        return addBigInt(i, 1, forceAppend);
    }

    public Integer addBytes32(String x, Boolean forceAppend) {
        String strWithoutPrefix = Numeric.cleanHexPrefix(x);
        if (strWithoutPrefix.length() > 64) {
            throw new IllegalArgumentException("invalid bytes32 str: too long" + x);
        }
        val strPadded = strWithoutPrefix + StringUtils.repeat("0", 64 - strWithoutPrefix.length());
        return insert(strPadded, forceAppend);
    }

    public Integer addHex(String x, Boolean forceAppend) {
        return insert(Numeric.cleanHexPrefix(x), forceAppend);
    }

    public Integer addRawBytes(byte[] bytes, Boolean forceAppend) {
        return insert(Numeric.cleanHexPrefix(Numeric.toHexString(bytes)), forceAppend);
    }

    public Integer addBigInt(BigInteger num, Integer numBytes, Boolean forceAppend) {
        return insert(Numeric.toHexStringNoPrefixZeroPadded(num, numBytes * 2), forceAppend);
    }

    private Integer insert(String x, Boolean forceAppend) {
        int offset = getLength();
        if (!forceAppend) {
            int start = 0;
            while (start != -1) {
                start = data.indexOf(x, start);
                if (start != -1) {
                    if ((start % 2) == 0) {
                        offset = start / 2;
                        return offset;
                    }
                }
            }
        }
        data += x;
        return offset;
    }
}

File: loopr-ios/Models/Sign/BitStream.swift
//
// Created by kenshin on 2019-04-04.
// Copyright (c) 2019 Loopring. All rights reserved.
//

import Foundation
import BigInt
import web3

class BitStream {

    let ADDRESS_LENGTH = 20
    let Uint256Max = BigInt(String(repeating: "f", count: 64), radix: 16)!

    var data: String = ""

    init() {

    }

    init(data: String) {
        self.data = data
    }

    func getData() -> String {
        return data.isEmpty ? "0x0" : "0x" + data;
    }

    func getBytes() -> [UInt8] {
        return data.hexBytes
    }

    func getLength() -> Int {
        return data.count / 2
    }

    func addAddress(_ x: String, forceAppend: Bool = true) {
        addAddress(x, ADDRESS_LENGTH, forceAppend: forceAppend);
    }
    
    func addAddress(_ x: String, _ numBytes: Int, forceAppend: Bool = true) {
        let _x = x.count == 0 ? "0" : x
        insert(_x.toBigInt!.toHexWithoutPrefixZero(size: numBytes * 2)!, forceAppend)
    }

    func addUint16(_ num: BigInt, forceAppend: Bool = true) {
        addBigInt(num, 2, forceAppend)
    }

    func addInt16(_ num: BigInt, forceAppend: Bool = true) {
        if num.signum() == -1 {
            let negUint256 = Uint256Max + num + 1
            let int16Str = negUint256.toHexWithoutPrefix()[60..<64]
            addHex(int16Str, forceAppend: forceAppend)
        } else {
            addBigInt(num, 2, forceAppend)
        }
    }

    func addUint32(_ num: BigInt, forceAppend: Bool = true) {
        addBigInt(num, 4, forceAppend)
    }

    func addUint(_ num: BigInt, forceAppend: Bool = true) {
        addBigInt(num, 32, forceAppend)
    }

    func addNumber(_ num: BigInt, _ numBytes: Int, forceAppend: Bool = true) {
        addBigInt(num, numBytes, forceAppend)
    }

    func addBool(_ b: Bool, forceAppend: Bool = true) {
        let _b = b ? BigInt(1) : BigInt(0)
        addBigInt(_b, 1, forceAppend)
    }

    func addBytes32(_ x: String, forceAppend: Bool) {
        let strWithoutPrefix = x.drop0x()
        if strWithoutPrefix.count <= 64 {
            let strPadded = strWithoutPrefix + String(repeating: "0", count: 64 - strWithoutPrefix.count)
            insert(strPadded, forceAppend)
        }
    }

    func addHex(_ x: String, forceAppend: Bool = true) {
        insert(x.drop0x(), forceAppend)
    }

    func addRawBytes(_ data: Data, forceAppend: Bool = true) {
        insert(data.hexStringWithNoPrefix, forceAppend)
    }

    func addBigInt(_ num: BigInt, _ numBytes: Int, _ forceAppend: Bool = true) {
        let x = num.toHexWithoutPrefixZero(size: numBytes * 2) ?? ""
        insert(x, forceAppend)
    }

    private func insert(_ x: String, _ forceAppend: Bool) {
        if !forceAppend {
            var start = 0
            while true {
                let subData = data.substring(fromIndex: start)
                if let index = subData.index(of: x) {
                    let distance = data.distance(from: data.startIndex, to: index)
                    if (distance % 2 == 0) {
                        return
                    } else {
                        start += 1
                    }
                } else {
                    break
                }
            }
        }
        data.append(x)
    }
}
